### PR TITLE
upload the improved boundary condition

### PR DESCRIPTION
**Pull request by ShuoguoZhangTUM**
upload the improved boundary condition

### CODE REVIEW

**Comment by Xiangyu-Hu** [comment]
@ShuoguoZhangTUM could you give itemized explanation on the changes? Thanks.

**Comment by ShuoguoZhangTUM** [comment]
In this version, I modify the pressure boundary file and the bidirectional buffer file
For the pressure boundary, now I use the template to define the velocity profile, and the bodyalignedbox to define the buffer region.
For the bidirectional buffer, now I use the template to prescribe pressure on populating particles. Here, I consider two cases: the bidirectional buffer works with pressure boundary and velocity boundary, respectively.

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h [hunk 1]
```diff
@@ -21,10 +21,10 @@
...
- * @file 	fluid_boundary.h
- * @brief 	Here, we define the boundary condition classes for fluid dynamics.
+ * @file 	pressure_boundary.h
+ * @brief 	Here, we define the pressure boundary condition class for fluid dynamics.
  * @details The boundary conditions very often based on different types of buffers.
- * @author	Chi Zhang and Xiangyu Hu
+ * @author  Shuoguo Zhang and Xiangyu Hu
  */
 
 #ifndef Pressure_BOUNDARY_H
```
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h:30]: shoudl be #ifndef PRESSURE_BOUNDARY_H

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h [hunk 2]
```diff
@@ -36,23 +36,34 @@ namespace SPH
...
   public:
-    FlowPressureBuffer(BodyPartByCell &body_part, Vecd normal_vector);
-    virtual ~FlowPressureBuffer(){};
-    void update(size_t index_i, Real dt = 0.0);
+    /** default parameter indicates prescribe pressure */
+    explicit PressureCondition(BodyAlignedBoxByCell &aligned_box_part)
+        : BaseFlowBoundaryCondition(aligned_box_part),
+          aligned_box_(aligned_box_part.aligned_box_),
+          transform_(aligned_box_.getTransform()), 
+          target_pressure(*this),
```
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h:48]: target_pressure_;

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h [hunk 3]
```diff
@@ -36,23 +36,34 @@ namespace SPH
...
 
   protected:
-    StdLargeVec<Vecd> &kernel_sum_;   
-    Vecd direction_;
-    /** Profile to be defined in applications,
-     * argument parameters and return value are in frame (local) coordinate */
-    virtual Real getTargetPressure(Real dt) = 0;
+    AlignedBoxShape &aligned_box_;
+    Transform &transform_;
+    TargetPressure target_pressure;
```
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h:65]: target_pressure_

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.cpp
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.cpp:1]: this file can be deleted?

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h [hunk 4]
```diff
@@ -38,118 +37,118 @@ namespace SPH
...
+    }
+};
+
+template <typename TargetPressure>
 class BidirectionalBuffer
 {
   protected:
-    ConcurrentIndexVector buffer_particle_list_;
-    virtual Real getTargetPressure(Real dt) = 0;
+    TargetPressure target_pressure;
```
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h:55]: target_pressure_;

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h [hunk 5]
```diff
@@ -38,118 +37,118 @@ namespace SPH
...
+                        size_t body_buffer_width) : target_pressure(*this),                                                                                                  
+        tag_buffer_particles(aligned_box_part, axis_direction),                                                                                                        
+        injection(aligned_box_part, axis_direction, body_buffer_width, *this){};
     virtual ~BidirectionalBuffer(){};
-    
-    TagBufferParticles<execution::ParallelPolicy> tag_buffer_particles;
-    Injection<execution::ParallelPolicy> injection;   
+
+    SimpleDynamics<TagBufferParticles> tag_buffer_particles;
+    SimpleDynamics<Injection> injection;
```
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h:148]: tag_buffer_particles_;
injection_

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.cpp
Comment (DrChiZhang) [inline tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.cpp:1]: this file can be deleted?

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h (hunk 1 above)
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h:30]: already done

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h (hunk 2 above)
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h:48]: already done

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h (hunk 3 above)
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.h:65]: already done

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.cpp
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/pressure_boundary.cpp:1]: yes, the details of function update have been merged into .h file

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h (hunk 4 above)
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h:55]: already done

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h (hunk 5 above)
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.h:148]: already done

File: tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.cpp
Comment (ShuoguoZhangTUM) [reply tests/extra_source_and_tests/extra_src/shared/pressure_boundary/bidirectional_buffer.cpp:1]: yes, , its deleted. the details of function update have been merged into .h file

**Review by Xiangyu-Hu — APPROVED** [review]
